core/app/models/concerns/spree/default_price.rb

```diff
@@ -5,20 +5,54 @@ module DefaultPrice
     extend ActiveSupport::Concern
 
     included do
-      has_one :default_price,
-        -> { with_discarded.currently_valid.with_default_attributes },
-        class_name: 'Spree::Price',
-        inverse_of: :variant,
-        dependent: :destroy,
-        autosave: true
+      delegate :display_price, :display_amount, :price, to: :default_price, allow_nil: true
+      delegate :price=, to: :default_price_or_build
+
+      # @see Spree::Variant::PricingOptions.default_price_attributes
+      def self.default_pricing
+        Spree::Config.default_pricing_options.desired_attributes
+      end
     end
 
-    def find_or_build_default_price
-      default_price || build_default_price(Spree::Config.default_pricing_options.desired_attributes)
+    # Returns {#default_price} or builds it from {Spree::Price.default_pricing}
+    #
+    # @return [Spree::Price, nil]
+    # @see Spree::Price.default_pricing
+    def default_price_or_build
+      default_price ||
+        prices.build(self.class.default_pricing)
     end
 
-    delegate :display_price, :display_amount, :price, to: :find_or_build_default_price
-    delegate :price=, to: :find_or_build_default_price
+    # Select from {#prices} the one to be considered as the default
+    #
+    # This method works with the in-memory association, so non-persisted prices
+    # are taken into account. Discarded prices are also considered.
+    #
+    # A price is a candidate to be considered as the default when it meets
+    # {Spree::Variant.default_pricing} criteria. When more than one candidate is
+    # found, non-persisted records take preference. When more than one persisted
+    # candidate exists, the one most recently updated is taken or, in case of
+    # race condition, the one with higher id.
+    #
+    # @return [Spree::Price, nil]
+    # @see Spree::Price.default_pricing
+    def default_price
```

Comment:
This method is a bit complex and hard to read to understand what it does. Any thoughts about moving it to a separate class that is responsible to select the default price only?

Reply:
Yeah, I completely agree. I have some reserves because of:

- It would be nice to have it be injectable for the end-user. However, we haven't decided how to do that yet. We could build on top of the old system, though.
- The logic around prices is very complex, and I'd like to simplify it further. However, probably it's not a preference at this point.

WDYT? We can do it now if you think it's better to leave it clean at this point.

Reply:
As we discussed IRL yesterday. Maybe a good tradeoff could be trying to use more semantic variable names and/or extracting something to a method in this same class to help reading the code.

Reply:
I tried to make everything clear. I think that the result is much more readable. Thanks for pointing it out. Please, take a look when you have a second.

Reply:
I think this is perfect, thanks Marc!